src/s3-copy.py

```diff
@@ -62,26 +66,46 @@
     # construct source filename
     fn_template = source["file_template"]["prefix"]+source["file_template"]["root"]+source["file_template"]["suffix"]
     source_dict = {"Bucket" : source["bucket"], "Key" : current_cycle.strftime(source["key"]+fn_template)}
-    # if source file exists, then copy
+
+    # check first if destination exists and it is the same as source
+    try: md5destination = destination_bucket.Object(destination_key).e_tag[1 :-1]
+    except: pass
+    else: #destination file exists
+      try: md5source = source_buckets[si].Object(source_dict["Key"]).e_tag[1 :-1]
+      except: pass
+      else: #source file exists
+        if (md5source == md5destination):
+          # log that this file is present 
+          logger_present.write("s3://{}/{} == s3://{}/{}\n".format(source_dict["Bucket"], source_dict["Key"], destination["bucket"], destination_key))
+          print(f" present {si}")
+          # skip to next, no need to copy identical file
+          break
+
+    # if source file exists but the destiantion doesnt, then copy
     try:
       destination_bucket.copy( source_dict, destination_key)
     except botocore.exceptions.ClientError as e:
       if e.response['Error']['Code'] == "404":
         # log that the file is missing and try next source
         logger_missing[si].write("s3://{}/{}\n".format(source_dict["Bucket"], source_dict["Key"]));
+        print(f" missing {si}")
         continue
       else:
         # something else went wrong
         raise
     else:
       # log success message
       logger_success.write("s3://{}/{} -> s3://{}/{}\n".format(source_dict["Bucket"], source_dict["Key"], destination["bucket"], destination_key))
+      print(f" copy from {si}")
+      break
+  # increment time counter
   dr.increment(seconds = cycling_interval)
 print('done with the time loop')
 
 #----- close logging files
 print('closing loggers')
 logger_success.close()
```

Comment:
For my own information, what is the difference between `success` and `present`? What do both imply?

Reply:
great question. 
`success` means the file was successfully copied. `present` means the file was already there and matched the source file. 
see examples of my log files here:
`https://noaa-reanalyses-pds.s3.amazonaws.com/index.html#observations/reanalysis/conv/gefs-prepbufr/log/`

note that this file path will disappear in the next 30 minutes or so and reappear here:
`https://noaa-reanalyses-pds.s3.amazonaws.com/index.html#observations/reanalysis/conv/prepbufr/log/`

as we discussed, i am still trying to figure out what is the right way to keep track of the file movements. I am considering adding md5 sum to the end of the line for success and present messages.